### PR TITLE
Frama-C: new release (27.0-Cobalt)

### PR DESCRIPTION
**Pull request by AllanBlanchard**
Main changes

### Kernel
- Supports C11 `_Generic`
- New machdep mechanism, based on YAML files
- New script for generating machdeps from a C11 compiler

### Aoraï
- Supports specification about floating-point variables.

### Eva
- The octagon domain can now infer relations between any lvalues of integer or
  pointer types.
- Fixes the bitwise domain on big-endian architectures.

### WP
- Why3 version bumped to 1.6.0.
- New options for controlling goal automatic splitting
- Default timeout is now 2s

### GTK GUI
- Removed GTK2 support

### Ivette
- The Eva table can show the values of function parameters, the logical status
  of ACSL predicates, and the values of C lvalues in these predicates.
  It also shows the status of uninitialized and escaping variables.
- Information about C types are shown in the Inspector.
- Many bug fixes and user experience improvements.

### CODE REVIEW

**Comment by AllanBlanchard** [comment]
We fixed the bug but forgot to fix the test. Would a patch be OK? Else we'll have to make a release 27.1.

File: packages/frama-c/frama-c.27.0/opam [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+   "DOCDIR=%{doc}%" { with-doc }
+   "install"
+  ]
+]
+
+remove: [
+  [make "PREFIX=%{prefix}%" "-f" "ivette/Makefile.installation" "uninstall"]
+]
+
+run-test: [
```
Comment (avsm) [inline packages/frama-c/frama-c.27.0/opam:109]: You can just delete this test clause if you want to "fix" the failed tests without a new release. There's no obligation in opam-repository to have tests in packages.

File: packages/frama-c/frama-c.27.0/opam [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  ["dune" "build" "-j%{jobs}%" "@ptests_config"
+  ] { arch != "ppc64" & arch != "x86_32" & arch != "arm32" & os != "macos" }
+]
+
+depends: [
+  "dune" { >= "3.2.0" | (>= "3.5.0" & os="macos") }
+  "dune-configurator"
+  "dune-site"
+
+  ( "alt-ergo-free" | "alt-ergo" )
```
Comment (avsm) [inline packages/frama-c/frama-c.27.0/opam:121]: I think you need a minimum bound on `alt-ergo-free` (see the failing "lower bounds" test in the CI)

File: packages/frama-c/frama-c.27.0/opam (hunk 1 above)
Comment (AllanBlanchard) [reply packages/frama-c/frama-c.27.0/opam:109]: I'll discuss this proposition with the team.

File: packages/frama-c/frama-c.27.0/opam (hunk 2 above)
Comment (AllanBlanchard) [reply packages/frama-c/frama-c.27.0/opam:121]: I believe that the problem is in the alt-ergo-free.2.0.0 opam package itself. In the same docker, when I try to install alt-ergo-free.2.0.0 (and nothing from Frama-C dependencies), I get the same error. 

Is it some kind of trap to convince me that patches are bad? :laughing: 

It is like something strange happens with [this patch](https://github.com/ocaml/opam-repository/blob/master/packages/alt-ergo-free/alt-ergo-free.2.0.0/files/patch_for_changes_in_num_library.patch). It is supposed to add required instructions to detect `num`, but while the compilation instruction indeed has a `@NUMLIB@` to replace:

```
ocamlopt.opt -c -annot -absname -bin-annot -short-paths -strict-sequence -g -inline 100 @NUMLIB@ -I /home/opam/.opam/4.13/lib/zarith ...
```

It seems that during configure, something is missing. Opam generates the following log:

```
# checking for ocamlfind... yes
# ocamlfind found zarith in -I /home/opam/.opam/4.13/lib/zarith
# ocamlfind found camlzip in -I /home/opam/.opam/4.13/lib/camlzip/../zip
# ocamlfind found ocplib-simplex in -I /home/opam/.opam/4.13/lib/ocplib-simplex
```

While when I configure by hand after applying the patch, I get:

```
checking for ocamlfind... yes
ocamlfind found zarith in -I /home/opam/.opam/4.13/lib/zarith
ocamlfind found num in -I /home/opam/.opam/4.13/lib/num
ocamlfind found camlzip in -I /home/opam/.opam/4.13/lib/camlzip/../zip
ocamlfind found ocplib-simplex in -I /home/opam/.opam/4.13/lib/ocplib-simplex
```

File: packages/frama-c/frama-c.27.0/opam (hunk 1 above)
Comment (AllanBlanchard) [reply packages/frama-c/frama-c.27.0/opam:109]: We simply remove the problematic test before the run.

File: packages/frama-c/frama-c.27.0/opam (hunk 2 above)
Comment (AllanBlanchard) [reply packages/frama-c/frama-c.27.0/opam:122]: Ok I don't know what happened but now the build of Alt-Ergo succeeds :thinking:

**Comment by avsm** [comment]
Many thanks for all the fixes @AllanBlanchard! I'll look into that strange `num` issue later but that's not a blocker to making your package available.

You may want to announce this on <https://discuss.ocaml.org>, where we have a Community category and an `announce` tag for this purpose.

**Pull request merged**